lib/lcdbwf/R/contrasts.R

```diff
@@ -154,85 +157,117 @@ make_results <- function(dds_name, label, dds_list=NULL, ...){
+      dots$test <- 'LRT'
+      test_detected <- TRUE
+    } else if (any(grepl('Wald', names(S4Vectors::mcols(dds))))) {
+      dots$test <- 'Wald'
+      test_detected <- TRUE
+    } else {
+      stop("test type was missing from make_results call and could not be detected from dds")
+    }
+  }
+
+  # Set the current default for 'type' from DESeq2 for lfcShrink if 'type' was not provided.
+  # This inspects the function definition of lfcShrink to see what the current default is
+  # (we have have seen it change before, hence the check).
+  if (!'type' %in% names(dots)) {
+    dots$type <- eval(formals(DESeq2::lfcShrink)$type)[1]
+  }
+
   # Call results() with the subset of dots that it accepts.
-  results_dots <- lcdbwf:::match_from_dots(dots, results)
-  res <- do.call("results", results_dots)
-
-  # We're about to call lfcShrink, but it needs the res object...so inject the
-  # one we just made into dots.
-  dots[['res']] <- res
-
-  # lfcShrink also needs the dds object, so inject that too
-  dots[['dds']] <- dds
-
-  lfcShrink_dots <- lcdbwf:::match_from_dots(dots, lfcShrink)
-    res <- do.call("lfcShrink", lfcShrink_dots)
-
-  # Add the shrinkage type to the metadata of the results object.
-  #
-  # If "type" was specified when calling this function, it's easy and we use
-  # that. Otherwise, if it was not specified then DESeq2 used the default.
-  # Since that default can change as we have seen in the past, we need to
-  # inspect the lfcShrink function itself to see what the current default is,
-  # and use that.
-  shrinkage_type <- dots[['type']]
-  if (is.null(shrinkage_type)){
-    # The definition of lfcShrink has a character vector as the type argument,
-    # and we want to extract the first thing in that vector. But formals()
-    # return strings, so we need to eval that string to convert it to
-    # a character vector such that we can extract the first thing.
-    #
-    # In recent versions this should evaluate to "apeglm". But this way we
-    # are inspecting the function itself if it ever changes.
-    shrinkage_type <- eval(formals(DESeq2::lfcShrink)$type)[1]
+  results_dots <- lcdbwf:::match_from_dots(dots, DESeq2::results)
+  res <- do.call(DESeq2::results, results_dots)
+
+  # When make_results is called with 'test' set to 'LRT',
+  # or when make_results is called with 'test' missing but
+  # DDS object contains the LRT, we convert all values in the log2FoldChange
+  # column of the DESeqResults object to 0. LFC values only make sense to report for a single
+  # comparison of two sample groups. This applies to the Wald test only.
+  # LRT is instead performing a test of the removal of one or more factor(s) from the design formula.
+  # DESeq2 reports log2FoldChange values for a single pair-wise comparison when test == 'LRT'. This
+  # can be misleading and so this is our solution.
+
+  # Adjust log2FoldChange for LRT test
+  if (!is.null(dots$test) && dots$test == 'LRT') {
+    res$log2FoldChange <- 0
```

Comment:
Add a warning that gets printed describing what happened, since this may be a surprising side-effect of running this function.